include/aws/io/host_resolver.h

```diff
@@ -211,6 +223,84 @@ AWS_IO_API size_t aws_host_resolver_get_host_address_count(
+    void (*release)(struct aws_host_resolver_listener *listener);
+};
+
+struct aws_host_resolver_listener {
+    struct aws_host_resolver_listener_vtable *vtable;
+    void *impl;
+};
+
+/* Callback for receiving a new host addresses from a listener. Memory for host_address is only guaranteed to exist
+ * during the callback, and must be copied if the caller needs it to persist after. */
+typedef void(aws_host_resolver_resolved_address_fn)(
+    struct aws_host_resolver_listener *listener,
+    struct aws_host_address *host_address,
+    void *user_data);
+
+typedef void(aws_host_resolver_listener_shutdown_fn)(void *user_data);
+
+struct aws_host_resolver_listener_options {
+
+    /* Name of the host to listen for notifications from. */
+    const struct aws_string *host_name;
+
+    /* Callback for when an address is resolved for the specified host. */
+    aws_host_resolver_resolved_address_fn *resolved_address_callback;
+
+    /* Callback for when a listener has completely shutdown. */
+    aws_host_resolver_listener_shutdown_fn *shutdown_callback;
+
+    /* User data to be passed into each callback. */
+    void *user_data;
+};
+
+/* Create and add a listener to the host resolver using the specified options. */
+AWS_IO_API struct aws_host_resolver_listener *aws_host_resolver_add_listener(
+    struct aws_host_resolver *resolver,
+    const struct aws_host_resolver_listener_options *options);
+
+/* Acquire a new reference to the listener. */
+AWS_IO_API void aws_host_resolver_listener_acquire(struct aws_host_resolver_listener *listener);
+
+/* Releases a reference to the listener. Once all references are released, the listener is removed from the host
+ * resolver. */
+AWS_IO_API void aws_host_resolver_listener_release(struct aws_host_resolver_listener *listener);
+
+/* Callback for receiving existing host addresses via aws_host_resolver_get_cached_addresses. Memory for host_address is
+ * only guaranteed to exist during the callback, and must be copied if the caller needs it to persist after. */
+typedef void(
+    aws_host_resolver_get_cached_addresses_callback_fn)(struct aws_host_address *host_address, void *user_data);
+
+/* Options structure for getting existing addresses of a host. */
+struct aws_host_resolver_get_cached_addresses_options {
+    struct aws_string *host_name;
+
+    /* Number of a-addresses to immediately retrieve from the cache. If the cache has less, it'll retrieve as
+     * many as possible. If set to g_aws_host_resolver_all_addresses, all available a-addresses will be returned.*/
+    size_t desired_num_a_addresses;
+
+    /* Number of aaaa-addresses to immediately retrieve from the cache. If the cache has less, it'll retrieve as
+     * many as possible. If set to g_aws_host_resolver_all_addresses, all available a-addresses will be returned.*/
+    size_t desired_num_aaaa_addresses;
+
+    aws_host_resolver_get_cached_addresses_callback_fn *get_cached_addresses_callback;
```

Comment:
I'm confused about how you're using all this stuff.
Are you getting cached addresses at the same time you're registering a listener, so that you can keep a mirror of all available addresses?

If so, don't also need to know when the addresses are evicted from the host-resolver?

Also, there's a gap where new addresses might be registered between getting the cached addresses and registering the listener. Or if you register the listener first then get cached you could end up with duplicates.

But really ... just why? Why mirror the host resolver at all? What about the host resolver isn't fast enough for you that you need to duplicate all its data?

Reply:
It's not a speed problem--it's a distribution problem.  The only available mechanism for querying an address from the host resolver is to get the next LRU address.  I think that's probably what you want 98% of the time, so that makes sense.  But for this particular use case, we want to divide traffic such that we have x number of resolved addresses, with y connections assigned to each of those resolved addresses.  Only being able to ask the host resolver for the next LRU address makes that difficult, because we only want to spin up the next set of connections when we have a new address. :(  It's also awkward because we want to share the host resolver to cut down on cost (ie: other people could be affecting what the LRU address is, someone else may have set up the host configuration).

We can totally add an on-evict callback in the same place.  It's not something I anticipate needing right now.  In trying to keep this change smaller, as long as we had a place to put it, I was okay with waiting until the need arose for it.  One reason I'm not using it is because the first caller to resolve the host gets to set the TTL.  If that's not my system for some reason, I don't control that value.  Another reason is because TTL may be a bit artificial/arbitrary for this use case.  We may abandon traffic in the future on certain addresses, but I'm skeptical that TTL would be the driving factor, or at the very least, the only factor.

In short, I don't need an exact mirror, I'm just looking to get x number of resolved addresses so that I can use them in a different way.  If they're available, I want to get them immediately.  If they're not, I want to wait until they show up.

And you're right, get-cache and add-listener calls are not guaranteed to be in sync/atomic.  I was worried that might be overkill, and was thinking that initially checking for duplicates or occasionally having an initial missed address in between those two calls would actually be okay.  But you're right, that's lame--I'll try to fix that.